lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java

```diff
@@ -16,28 +16,185 @@
+            log.error("Unexpected exception while creating client results", e);
+            unlock(response);
+            throw e;
+        }
     }
 
     @Override
     public void close() {
         // TODO(aalouane): decide whether or not to close the unlocker depending on ownership
     }
+
+    private TimestampLeaseResponses acquireLeasesWithRetry(Map<TimestampLeaseName, Integer> requests) {
+        try {
+            return retryer.call(() -> acquireLeases(requests)).orElseThrow();
+        } catch (ExecutionException e) {
+            // should not happen with how the retryer is set up
+            log.warn("Unexpected exception. Expected a retry exception", e);
+            throw new SafeRuntimeException(e);
+        } catch (RetryException e) {
+            Attempt<?> lastFailedAttempt = e.getLastFailedAttempt();
+            log.warn(
+                    "Exhausted retries to get enough timestamps while acquiring timestamp lease",
+                    SafeArg.of("requests", requests),
+                    SafeArg.of("numRetries", lastFailedAttempt.getAttemptNumber()),
+                    SafeArg.of("attemptHadException", lastFailedAttempt.hasException()),
+                    e);
+            throw new AtlasDbDependencyException(
+                    "Exhausted retries to get enough timestamps while acquiring timestamp lease",
+                    lastFailedAttempt.hasException() ? lastFailedAttempt.getExceptionCause() : null,
+                    SafeArg.of("requests", requests),
+                    SafeArg.of("numRetries", lastFailedAttempt.getAttemptNumber()),
+                    SafeArg.of(
+                            "maybeResult",
+                            lastFailedAttempt.hasResult() ? lastFailedAttempt.getResult() : Optional.empty()));
+        }
+    }
+
+    private Optional<TimestampLeaseResponses> acquireLeases(Map<TimestampLeaseName, Integer> requestedFreshTimestamps) {
+        // we prefer to use a new request id for non-dialogue-native dialogue attempts
+        TimestampLeaseRequests request =
+                TimestampLeaseRequests.of(RequestId.of(uuidSupplier.get()), requestedFreshTimestamps);
+        NamespaceTimestampLeaseRequest requests = NamespaceTimestampLeaseRequest.of(List.of(request));
+
+        List<TimestampLeaseResponses> responseList =
+                delegate.acquireTimestampLeases(requests).get();
+        if (responseList.size() != 1) {
+            responseList.forEach(this::unlock);
+            throw new SafeRuntimeException(
+                    "Response should include exactly one result",
+                    SafeArg.of("request", request),
+                    SafeArg.of("responseList", responseList));
+        }
+
+        TimestampLeaseResponses response = responseList.get(0);
+        Map<TimestampLeaseName, TimestampLeaseResponse> responseMap = response.getTimestampLeaseResponses();
+        if (!requestedFreshTimestamps.keySet().equals(responseMap.keySet())) {
```

Comment:
Again I'm kind of ok to drop this on the floor

Reply:
Ok let's go

Reply:
Yeah, like timelock should not be doing this.